src/particles/elements/NonlinearLens.H

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2022 Chad Mitchell, Axel Huebl
+ *
+ * This file is part of ImpactX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef IMPACTX_NONLINEARLENS_H
+#define IMPACTX_NONLINEARLENS_H
+
+#include "particles/ImpactXParticleContainer.H"
+
+#include <AMReX_Extension.H>
+#include <AMReX_REAL.H>
+#include <AMReX_GpuComplex.H>
+
+#include <cmath>
+
+namespace impactx
+{
+    struct NonlinearLens
+    {
+        using PType = ImpactXParticleContainer::ParticleType;
+
+        /** A thin lens associated with a single short segment of the
+         *  nonlinear magnetic insert described by V. Danilov and
+         *  S. Nagaitsev, PRSTAB 13, 084002 (2010), Sect. V.A.  This
+         *  element appears in MAD-X as type NLLENS.
+         *
+         * @param knll - integrated strength of the nonlinear lens (m)
+         * @param cnll - distance of singularities from the origin (m)
+         *
+         */
+        NonlinearLens( amrex::ParticleReal const knll,
+                   amrex::ParticleReal const cnll )
+        : m_knll(knll), m_cnll(cnll)
+        {
+        }
+
+        /** This is a nonlinear lens functor, so that a variable of this type can be used like a
+         *  nonlinear lens function.
+         *
+         * @param p Particle AoS data for positions and cpu/id
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         * @param pt particle momentum in t
+         * @param refpart reference particle (unused)
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void operator() (
+                PType& AMREX_RESTRICT p,
+                amrex::ParticleReal & AMREX_RESTRICT px,
+                amrex::ParticleReal & AMREX_RESTRICT py,
+                amrex::ParticleReal & AMREX_RESTRICT pt,
+                [[maybe_unused]] RefPart const refpart) const {
+
+            using namespace amrex::literals; // for _rt and _prt
```

Comment:
Oh, actually here and in #118 we need to do:
```suggestion
            using namespace amrex::literals; // for _rt and _prt
            
            using Complex = amrex::GpuComplex<amrex::ParticleReal>;
```
to define the type of the currently compiled precision. Otherwise it defaults to `double`.

Then below...

Reply:
Ok, I see.  Thanks for making this change here and in #118.